File: Online_Judge_Problems/boj/boj_16497_대출요청/boj_16497.js
/**
 * -----------------------------------------------------------
 * Sub    : [BOJ] 대출 요청 (16497)
 * Link   : https://www.acmicpc.net/problem/16497
 * Level  : Bronze 2
 * Tag    : JavaScript (Node.js)
 * ------------------------------------------------------------
 */

const fs = require("fs");

function solve() {
  // 1. 입력 처리: 표준 입력(0)으로부터 전체 데이터를 읽어와 공백 단위로 분리합니다.
  const input = fs.readFileSync(0, "utf-8").trim().split(/\s+/).map(Number);

  let idx = 0;
  const N = input[idx++]; // 대출 요청 건수

  const requests = [];
  for (let i = 0; i < N; i++) {
    const s = input[idx++];
    const e = input[idx++];
    requests.push([s, e]);
  }

  const K = input[idx++]; // 보유한 책의 권수

  // 2. 날짜별 대출 현황 계산 (1일 ~ 31일)
  // Array.fill을 이용해 0으로 초기화된 배열을 생성합니다.
  const days = new Array(32).fill(0);

  for (const [start, end] of requests) {
    // 반납일 당일(end)은 대출 가능 상태이므로 end 전까지만 카운트합니다.
    for (let i = start; i < end; i++) {
      days[i]++;
    }
  }

  // 3. 결과 판별 및 출력
  // Math.max(...array)를 사용하여 배열의 최댓값을 구합니다.
  const maxRequired = Math.max(...days);

  console.log(maxRequired > K ? 0 : 1);
}

solve();

File: boj/boj_16497_대출요청/boj_16497.py
"""
-----------------------------------------------------------
Sub    : [BOJ] 대출 요청
Link   : https://www.acmicpc.net/problem/16497
Level  : 브론즈 2
Tag    : Python
-----------------------------------------------------------
Details

-----------------------------------------------------------
"""

import os
import sys


def get_input():
    if not sys.stdin.isatty():
        f = sys.stdin
    else:
        file_path = os.path.join(os.path.dirname(__file__), "input_test.txt")
        if os.path.exists(file_path):
            f = open(file_path, "r", encoding="utf-8")
        else:
            f = sys.stdin

    for line in f:
        for char in line.split():
            yield char


tokens = get_input()


def input_token():
    return next(tokens)


def solve():
    try:
        n_str = input_token()
        if not n_str:
            return
        n = int(n_str)

        requests = []
        for _ in range(n):
            s = int(input_token())
            e = int(input_token())
            requests.append((s, e))

        k = int(input_token())

        days = [0] * 32
        for s, e in requests:
            for i in range(s, e):
                days[i] += 1

        if max(days) > k:
            print(0)
        else:
            print(1)

    except (StopIteration, ValueError):
        pass


if __name__ == "__main__":
    solve()
